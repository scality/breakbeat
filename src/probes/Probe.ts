export interface Probe {
    check(): void;
    get value(): boolean;
}
