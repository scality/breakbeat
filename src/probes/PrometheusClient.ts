import * as joi from '@hapi/joi';
import 'joi-extract-type';

import { PrometheusDriver } from 'prometheus-query';
import { Logger } from 'werelogs';

const log = new Logger('breakbeat:prometheusClient');

export const prometheusClientConfigSchema = joi.object({
    endpoint: joi.string().uri({
        scheme: ['http', 'https'],
        allowRelative: false,
    }).required(),
    timeout: joi.number().optional(),
});

export type PrometheusClientConfig = joi.extractType<typeof prometheusClientConfigSchema>;

type InstantResult = {
    value: {
        value: number,
    },
};

export class PrometheusClient {
    prom: PrometheusDriver;
    q: string;

    constructor(config: PrometheusClientConfig, q: string, interval?: string) {
        this.prom = new PrometheusDriver({
            endpoint: config.endpoint,
            timeout:  config.timeout,
        });

        this.q = interval ?
            `sum(avg_over_time(${q}[${interval}]))` :
            `sum(${q})`;
    }

    async instantQuery(): Promise<number | null> {
        try {
            const res = await this.prom.instantQuery(this.q);

            // res.result is an any[]
            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
            const series: InstantResult[] = res.result;
            if (series.length == 0) {
                return null;
            }

            return series[0].value.value;
        } catch (error: unknown) {
            log.error('unable to query prometheus', { error });
            return null;
        }
    }
}
