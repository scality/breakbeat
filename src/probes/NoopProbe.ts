import * as joi from '@hapi/joi';

import { Probe } from './Probe';

export const noopProbeSchema = joi.object({
    type: joi.string().valid('noop').required(),
    returnConstantValue: joi.boolean().required(),
});

export type NoopProbeConfig = joi.extractType<typeof noopProbeSchema>;

export class NoopProbe implements Probe {
    config: NoopProbeConfig;

    constructor(config: NoopProbeConfig) {
        this.config = config;
    }

    check(): Promise<void> {
        return Promise.resolve();
    }

    get value() {
        return this.config.returnConstantValue;
    }
}
