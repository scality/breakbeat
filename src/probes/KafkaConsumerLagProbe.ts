import * as joi from '@hapi/joi';
import 'joi-extract-type';

import { Probe } from './Probe';
import { PrometheusClient, prometheusClientConfigSchema } from './PrometheusClient';

import { Logger } from 'werelogs';

const log = new Logger('breakbeat:probe:kafkaConsumerLag');

export const kafkaConsumerLagProbeSchema = joi.object({
    type: joi.string().valid('kafkaConsumerLag').required(),

    prometheus: prometheusClientConfigSchema.required(),

    wantTotalLagLessThan: joi.number().greater(0).required(),
    averagedOverInterval: joi.string().regex(new RegExp('^\\d+[smhd]$')).optional(),
    consumerGroupName: joi.string().required(),
    topicName: joi.string().optional(),
});

export type KafkaConsumerLagProbeConfig = joi.extractType<typeof kafkaConsumerLagProbeSchema>;

export class KafkaConsumerLagProbe implements Probe {
    config: KafkaConsumerLagProbeConfig;
    lag: number;
    prometheusClient: PrometheusClient;

    constructor(config: KafkaConsumerLagProbeConfig) {
        const topic = config.topicName ? `,topic="${config.topicName}"` : '';
        const q = `kafka_consumergroup_group_lag{
            group="${config.consumerGroupName}"
            ${topic}
        }`;

        this.prometheusClient = new PrometheusClient(
            config.prometheus, q, config.averagedOverInterval);
        this.config = config;
        this.lag = 0;
    }

    async check() {
        const v = await this.prometheusClient.instantQuery();
        if (v != null) {
            if (Number.isNaN(v)) {
                log.warn('warning: ignoring received NaN value (interval too long for retention?)');
                return;
            }
            this.lag = v;
        }
    }

    get value() {
        return this.lag < this.config.wantTotalLagLessThan;
    }
}
