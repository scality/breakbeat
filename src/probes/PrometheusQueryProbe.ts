import * as joi from '@hapi/joi';
import 'joi-extract-type';

import { Probe } from './Probe';
import { PrometheusClient, prometheusClientConfigSchema } from './PrometheusClient';

import { Logger } from 'werelogs';

const log = new Logger('breakbeat:probe:prometheusQuery');

export const prometheusQueryProbeSchema = joi.object({
    type: joi.string().valid('prometheusQuery').required(),

    prometheus: prometheusClientConfigSchema.required(),

    query: joi.string().required(),
    threshold: joi.number().required(),
    averagedOverInterval: joi.string().regex(new RegExp('^\\d+[smhd]$')).optional(),
});

export type PrometheusQueryProbeConfig = joi.extractType<typeof prometheusQueryProbeSchema>;

export class PrometheusQueryProbe implements Probe {
    config: PrometheusQueryProbeConfig;
    threshold: number;
    prometheusClient: PrometheusClient;
    observed: number;

    constructor(config: PrometheusQueryProbeConfig) {
        this.prometheusClient = new PrometheusClient(
            config.prometheus, config.query, config.averagedOverInterval);
        this.config = config;
        this.threshold = 0;
        this.observed = 0;
    }

    async check() {
        const v = await this.prometheusClient.instantQuery();
        if (v != null) {
            if (Number.isNaN(v)) {
                log.warn('warning: ignoring received NaN value (interval too long for retention?)');
                return;
            }
            this.observed = v;
        }
    }

    get value() {
        return this.observed < this.config.threshold;
    }
}
