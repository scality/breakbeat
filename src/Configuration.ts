import * as joi from '@hapi/joi';
import 'joi-extract-type';

import { probeSchema } from './probes';

export const defaultProbeEvaluateInterval = 60000;

const topSchema = joi.object({
    probes: joi.array().items(probeSchema).optional(),
    nominalEvaluateIntervalMs: joi.number().default(defaultProbeEvaluateInterval).optional(),
    trippedEvaluateIntervalMs: joi.number().default(defaultProbeEvaluateInterval).optional(),
    stabilizingEvaluateIntervalMs: joi.number().default(defaultProbeEvaluateInterval).optional(),
    stabilizeAfterNSuccesses: joi.number().default(2).optional(),
});

export type Config = joi.extractType<typeof topSchema>;

export function validate(config: unknown): Config {
    if (!config) {
        return validate({});
    }

    const res = topSchema.validate(config);
    if (res.error) {
        throw res.error;
    }

    return res.value;
}

export const configWithDefaults = validate({});
