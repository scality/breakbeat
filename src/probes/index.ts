import * as joi from '@hapi/joi';
import 'joi-extract-type';

import { NoopProbe, noopProbeSchema } from './NoopProbe';
import { KafkaConsumerLagProbe, kafkaConsumerLagProbeSchema } from './KafkaConsumerLagProbe';
import { Probe } from './Probe';

const probeSchema = joi.alternatives().try(
    noopProbeSchema.required(),
    kafkaConsumerLagProbeSchema.required(),
);

export type ProbeConfig = joi.extractType<typeof probeSchema>;

function buildProbe(probeConfig: ProbeConfig): Probe {
    const type = probeConfig.type;
    switch (type) {
    case 'noop':
        return new NoopProbe(probeConfig);
    case 'kafkaConsumerLag':
        return new KafkaConsumerLagProbe(probeConfig);
    }

    // not really dead code for transpiled uses
    const tpe: string = probeConfig['type'];
    throw new Error(`unsupported probe type '${tpe}'`);
}

export {
    Probe,
    buildProbe,
    probeSchema,
};
