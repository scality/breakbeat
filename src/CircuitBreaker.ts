import { Config, defaultProbeEvaluateInterval, validate } from './Configuration';
import { Probe, buildProbe } from './probes';

export enum BreakerState {
    Nominal = 0,
    Stabilizing = 1,
    Tripped = 2,
}

export class CircuitBreaker {
    _config: Config;
    _probes: Probe[];

    _evaluateIntervalByAggregateState: {
        [key in BreakerState]: number;
    };

    _aggregateState: BreakerState;
    _stabilizingCounter: number;
    _stabilizeThreshold: number;
    _evaluateTimeoutHandle: NodeJS.Timeout | null;

    // a promise that unit tests can await on, as jest's fake timers
    // get confused with async flows
    _evaluatingPromiseHook: Promise<unknown> | null;

    constructor(config: unknown) {
        this._config = validate(config);
        this._probes = (this._config.probes || []).map(buildProbe);

        this._aggregateState = BreakerState.Nominal;
        this._stabilizingCounter = 0;

        this._stabilizeThreshold = this._config.stabilizeAfterNSuccesses || 0;
        this._evaluateIntervalByAggregateState = {
            [BreakerState.Nominal]: this._config.nominalEvaluateIntervalMs || defaultProbeEvaluateInterval,
            [BreakerState.Stabilizing]: this._config.stabilizingEvaluateIntervalMs || defaultProbeEvaluateInterval,
            [BreakerState.Tripped]: this._config.trippedEvaluateIntervalMs || defaultProbeEvaluateInterval,
        };

        this._evaluateTimeoutHandle = null;
        this._evaluatingPromiseHook = null;
    }

    get state(): BreakerState {
        return this._aggregateState;
    }

    start() {
        this._aggregateState = BreakerState.Nominal;
        this._scheduleNextEvaluation();
    }

    stop() {
        if (this._evaluateTimeoutHandle) {
            clearTimeout(this._evaluateTimeoutHandle);
            this._evaluateTimeoutHandle = null;
        }
    }

    async _evaluate() {
        this._evaluatingPromiseHook = Promise.allSettled(this._probes.map(p => p.check()));
        await this._evaluatingPromiseHook;
        this._evaluatingPromiseHook = null;

        const allOk = this._probes.every(probe => probe.value);

        if (allOk) {
            switch (this._aggregateState) {
            case BreakerState.Tripped:
                this._aggregateState = BreakerState.Stabilizing;
                this._stabilizingCounter = 1;
                break;

            case BreakerState.Stabilizing:
                this._stabilizingCounter++;

                if (this._stabilizingCounter >= this._stabilizeThreshold) {
                    this._aggregateState = BreakerState.Nominal;
                }
                break;

            default:
                break;
            }
        } else {
            this._aggregateState = BreakerState.Tripped;
        }

        // only schedule next if start was called before
        if (this._evaluateTimeoutHandle) {
            this._scheduleNextEvaluation();
        }
    }

    _scheduleNextEvaluation() {
        this._evaluateTimeoutHandle = setTimeout(
            () => void this._evaluate(),
            this._evaluateIntervalMs,
        );
    }

    get _evaluateIntervalMs(): number {
        return this._evaluateIntervalByAggregateState[this._aggregateState];
    }
}
