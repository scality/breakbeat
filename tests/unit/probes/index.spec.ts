import { describe, expect, test } from '@jest/globals';

import { buildProbe } from '../../../src/probes';

import { NoopProbe } from '../../../src/probes/NoopProbe';
import { KafkaConsumerLagProbe } from '../../../src/probes/KafkaConsumerLagProbe';

describe('Probe', () => {
    describe('buildProbe', () => {
        describe('noop', () => {
            test('should build a noop probe', () => {
                const noop = buildProbe({
                    type: 'noop',
                    returnConstantValue: true,
                });

                expect(noop).toBeInstanceOf(NoopProbe);
            });
        });

        describe('kafkaConsumerLag', () => {
            test('should build a noop probe', () => {
                const kcl = buildProbe({
                    type: 'kafkaConsumerLag',
                });

                expect(kcl).toBeInstanceOf(KafkaConsumerLagProbe);
            });
        });

        test('should throw if unknown probe type', () => {
            const config = {
                type: 'unknown',
            };

            // Testing what would happen if transpiled code gave an
            // unsupported value, so disable type checking.
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            expect(() => buildProbe(config)).toThrowErrorMatchingSnapshot();
        });
    });
});
