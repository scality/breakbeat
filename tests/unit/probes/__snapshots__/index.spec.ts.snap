// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probe buildProbe should throw if unknown probe type 1`] = `"unsupported probe type 'unknown'"`;
